doc/dash/dash-acl-tags/dash-acl-tags.md

```diff
@@ -0,0 +1,336 @@
+
+### Configuration flow
+
+![Configuration flow](images/dash-acl-tags.png)
+
+### Detailed flow
+
+DashAclOrch is responsible for DASH ACL support and should be extended to support ACL tags. DashAclOrch subscribes to DASH_PREFIX_TAG_TABLE and DASH_ACL_RULE_TABLE updates from APP_DB, processes the configuration updates, and transforms them into SAI API calls.
+
+### Runtime
+
+When an update for the DASH_PREFIX_TAG_TABLE or DASH_ACL_RULE_TABLE table is received DashAclOrch will understand the operation type. The supported operation types are create, update, and remove.
+
+### Tags handling
+
+
+**Add new ACL tag**
+
+If the operation is to create a new ACL tag in DASH_PREFIX_TAG_TABLE DashAclOrch will do the following
+- Validate the tag configuration
+- Store the tag configuration
+
+**Remove ACL tag**
+
+If the operation is to remove the ACL tag from DASH_PREFIX_TAG_TABLE DashAclOrch will do the following
+- If the tag is attached to the rules
+    - Return status "retry" to postpone tag removal
+- Remove the tag
+
+**Update ACL tag**
+
+If the operation is to update the list of prefixes in the ACL tag in DASH_PREFIX_TAG_TABLE DashAclOrch will do the following
+  - Based on the list of ACL rules that uses the tag get list of affected ACL groups
+  - For each affected ACL group
+  - If group is bound to the ENI
+    - Create new ACL group
+    - For each ACL rule from the original ACL group
+      - Copy original ACL rule attributes
+      - Update attributes if required
+      - Create new ACL rule and attach it to the new ACL group
+    - Bind new ACL group to the ENI
+    - Remove all old ACL rules
+    - Remove old ACL group
+  - Else (if ACL group is not bound)
+    - For each ACL rule that uses the tag
+      - Update ACL rules
+
+### ACL rules handling
+
+
+**ACL rule creation flow**
+
+If the operation is to create a new ACL rule in DASH_ACL_RULE_TABLE DashAclOrch will do the following
+- If tags are specified in the configuration
+    - For each configured tag
+        - Check if the tags exists in `DASH_PREFIX_TAG_TABLE`. If the tag doesn't exist return status "retry" to postpone rule creation.
+    - If all tags are available
+      - Expand all tags to a list of prefixes
+      - Create ACL rule
+
+**Remove ACL rule**
+
+If the operation is to remove the existing ACL rule from DASH_ACL_RULE_TABLE DashAclOrch will do the following
+- Remove rule
+
+**Update ACL rule**
+
+If the operation is to update the existing ACL rule in DASH_ACL_RULE_TABLE DashAclOrch will do the following
+- if the ACL group to which the ACL rule belongs is bound to the ENI
+  - Create new ACL group
+  - For each ACL rule from the original ACL group
+    - Copy original ACL rule attributes
+    - Update attributes if required
+    - Create new ACL rule and attach it to the new ACL group
+  - Bind new ACL group to the ENI
+  - Remove all old ACL rules
+  - Remove old ACL group
```

Comment:
Why cannot we directly create a new ACL rule and delete the old ACL rule other than refreshing the whole acl group?

Reply:
When the ACL group is not bound to any ENIs we can do an update rule by rule, as this won't affect the traffic. 

But if the ACL group is bound to the ACL we want to perform update operation atomically. The tag may be used by more than one ACL rule in the group. If we update the rules one by one this will create inconsistency of the  ACL group. Some of the ACL rules will have updated prefixes and some of them not. To prevent inconsistency we need to refresh the entire ACL group.

Reply:
Yes, totally agree. Your mentioned is about updating **TAG entry**. But this section is about updating one **ACL rule**. I feel the logic will be simpler.

Reply:
Got you. I've just updated the `Remove ACL rule` and `Update ACL rule` sections to correctly cover this.
According to the ACL requirements in the [DASH SONiC HLD](https://github.com/sonic-net/DASH/blob/main/documentation/general/dash-sonic-hld.md#17-acl-requirements) document it is not allowed to update ACL rule from the ACL group that is bound to the ENI:

_It is not permitted to modify rules within a group that is currently bound to an ENI. For such scenario, a new group shall be created by application with the modified set of rules which could then be bind to ENI. An exception is if tags are expanded to individual prefixes in an ACL rule. In such cases, if tags are modified, application shall update the corresponding rule by adding/removing a prefix._

 In both cases, we will allow updating (or removing) the ACL rule only if the group is not bound to the ENI. Otherwise, an error will be removed.

If the updating or removal of the ACL rule from the ACL group that is bound to the ENI should be supported, we can use the same strategy that we have in the update tag flow:
- Create new ACL group
- For each ACL rule from the original ACL group
  - If the rule shall be updated: 
    - Update attributes
    - Create new ACL rule and attach it to the new ACL group
  - Else if the rule shall be removed
    - Continue (do not create rule)
  - Else
    - Create rule with no changes 
  - Bind new ACL group to the ENI
- Remove all old ACL rules
- Remove old ACL group

Reply:
@Pterosaur kindly reminder. Please review my comment and let me know if you agree.

Reply:
Hi, agree with your point.
But I'm just curious why it is not permitted to modify rules within a group? It's due to hardware limitations or just software implementation considerations?